### PR TITLE
fix concurrency bug in our concurrency library

### PR DESCRIPTION
**Pull request by clockfort**

### CODE REVIEW

**Comment by clockfort** [comment]
@rjullman and @carrino to review

File: commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java [hunk 1]
```diff
@@ -25,31 +31,48 @@
...
+    private final static ThreadLocal<LoadingCache<ExecutorInheritableThreadLocal<?>, Object>> perThreadThreadCache = new ThreadLocal<LoadingCache<ExecutorInheritableThreadLocal<?>, Object>>() {
         @Override
-        protected WeakHashMap<ExecutorInheritableThreadLocal<?>, Object> initialValue() {
-            return new WeakHashMap<ExecutorInheritableThreadLocal<?>, Object>();
+        protected LoadingCache<ExecutorInheritableThreadLocal<?>, Object> initialValue() {
+            return CacheBuilder.newBuilder().weakKeys().build(
+                    new CacheLoader<ExecutorInheritableThreadLocal<?>, Object>() {
+                        @Override
+                        public Object load(ExecutorInheritableThreadLocal<?> key) throws Exception {
+                            return new WeakHashMap<ExecutorInheritableThreadLocal<?>, Object>();
```
Comment (carrino) [inline commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java]: so we add another level of map here?

**Comment by carrino** [comment]
is this all just to allow nulls?

File: commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java [hunk 2]
```diff
@@ -25,31 +31,48 @@
  */
 public class ExecutorInheritableThreadLocal<T> {
 
-    private final static ThreadLocal<WeakHashMap<ExecutorInheritableThreadLocal<?>, Object>> mapForThisThread
-            = new ThreadLocal<WeakHashMap<ExecutorInheritableThreadLocal<?>,Object>>() {
+    private final static ThreadLocal<LoadingCache<ExecutorInheritableThreadLocal<?>, Object>> perThreadThreadCache = new ThreadLocal<LoadingCache<ExecutorInheritableThreadLocal<?>, Object>>() {
```
Comment (carrino) [inline commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java]: I think we may just want Cache and not LoadingCache

**Comment by carrino** [comment]
this weak map should only be accessed by the thread getting from the ThreadLocal.  This bug is bizarrro

**Comment by carrino** [comment]
I feel like if the bug is anywhere, it's in   this section         

```
 // Temporarily install the untransformed map in case callInstallOnChildThread makes use
            // of existing thread locals (UserSessionClientInfo does this).
            mapForThisThread.set(map);
            WeakHashMap<ExecutorInheritableThreadLocal<?>, Object> newMap = new WeakHashMap<ExecutorInheritableThreadLocal<?>, Object>(map);
            // Iterate over the new map so that 1) We modify entries in the new
            // map, not the old, and 2) so we don't get CMEs if
            // callInstallOnChildThread adds or removes any thread locals.
            for (Map.Entry<ExecutorInheritableThreadLocal<?>, Object> e : newMap.entrySet()) {
                e.setValue(e.getKey().callInstallOnChildThread(e.getValue()));
            }
```

File: commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java [hunk 3]
```diff
@@ -133,14 +156,14 @@ protected T initialValue() {
...
-        WeakHashMap<ExecutorInheritableThreadLocal<?>, Object> oldMap = mapForThisThread.get();
+    static ConcurrentMap<ExecutorInheritableThreadLocal<?>, Object> installMapOnThread(ConcurrentMap<ExecutorInheritableThreadLocal<?>, Object> map) {
+        ConcurrentMap<ExecutorInheritableThreadLocal<?>, Object> oldMap = perThreadThreadCache.get().asMap();
         if (map.isEmpty()) {
-            mapForThisThread.remove();
+            perThreadThreadCache.remove();
         } else {
             // Temporarily install the untransformed map in case callInstallOnChildThread makes use
             // of existing thread locals (UserSessionClientInfo does this).
-            mapForThisThread.set(map);
```
Comment (carrino) [inline commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java]: I think we can clone the map here and we should be good.

I'm also interested to find the code that causes this failure.  I expect it to live in a callInstallOnChildThread method

Comment (carrino) [inline commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java]: Or we can just make the incoming map Immutable and that will expose who is setting map stuff during callInstallOnChildThread

File: commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java [hunk 4]
```diff
@@ -133,14 +156,14 @@ protected T initialValue() {
...
+            perThreadThreadCache.get().asMap().putAll(map);
```
Comment (carrino) [inline commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java]: i think you need to invalidate the whole cache before putting here.

File: commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java [hunk 5]
```diff
@@ -116,13 +139,13 @@ protected T initialValue() {
...
+        ConcurrentMap<ExecutorInheritableThreadLocal<?>, Object> currentMap = perThreadThreadCache.get().asMap();
         if (currentMap.isEmpty()) {
-            mapForThisThread.remove();
+            perThreadThreadCache.remove();
             return currentMap;
         }
-        WeakHashMap<ExecutorInheritableThreadLocal<?>, Object> ret = new WeakHashMap<ExecutorInheritableThreadLocal<?>, Object>(currentMap.size());
+        ConcurrentMap<ExecutorInheritableThreadLocal<?>, Object> ret = new ConcurrentHashMap<ExecutorInheritableThreadLocal<?>, Object>(currentMap.size());
         for (Map.Entry<ExecutorInheritableThreadLocal<?>, Object> e : currentMap.entrySet()) {
             ret.put(e.getKey(), e.getKey().callChildValue(e.getValue()));
```
Comment (carrino) [inline commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java]: ok, the other possibility is that callChildValue does sets into the map for the current thread.

File: commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java (hunk 2 above)
Comment (carrino) [inline commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java]: actually, can we just have ConcurrentMap directly and avoid the cache

**Comment by carrino** [comment]
Ok, I don't think we ever use the Weak aspect because for the most part ExecutorInheritables are just static finals.  I think making everything a concurrent map is a good idea.

File: commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java (hunk 4 above)
Comment (clockfort) [inline commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java]: agree

**Comment by clockfort** [comment]
but https://jira.yojoe.local/browse/QA-75847 seems to suggest it'd leak memory otherwise?

File: commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java (hunk 5 above)
Comment (clockfort) [inline commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java]: I think overriding the method called here is how this person found the bug, where the bug is really that we handed someone a gun to break ThreadLocal-ness.

File: commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java (hunk 2 above)
Comment (clockfort) [inline commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java]: Cache is just for the weak keys

Comment (carrino) [inline commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java]: MapMaker.weakKeys()

**Comment by carrino** [comment]
Yes, it seems like concurrentmap will strictly be better

**Comment by schlosna** [comment]
@clockfort do you have any numbers for how this affects performance? I'm especially thinking of the default cached executor service with 100 ms TTL threads, where installing new maps is already quite expensive, I would expect CHM to be more so. Have you already looked at synchronizing over existing WHM (where I'd expect very little contention)?

**Comment by carrino** [comment]
Yeah, the other option is to disallow edits inside childValue and installOnChildThread

**Comment by carrino** [comment]
@schlosna Why does a low TTL cause a lot of map installs?  This code only runs at submit and execute time.  A thread being created or terminated shouldn't cause this code to run.

**Comment by schlosna** [comment]
@carrino The map is created per thread on first access to the `ThreadLocal`, so the low TTL leads to short thread lifetimes, which leads to more maps created, and more allocation/GC churn. 

CHM are much larger than WHM with all their lock striped segments. If we stick with CHM, we probably want to at least drop the concurrency level to 1 or 2 and let things contend in the worse case, and set a small initial capacity.

**Comment by carrino** [comment]
we can just use 1 because this map will only ever be accessed from one
thread

**Comment by clockfort** [comment]
closed for #293

**Pull request closed without merging**